### PR TITLE
Dataset concatenate

### LINKED ISSUE #387: Dataset concatenate
Add concatenate functionality to Dataset class, implement `concatenate`.

See https://github.com/scipp/scipp/tree/v0.0.1 and in particular dataset.h (before the refactor) for what was previously supported.

### PR DESCRIPTION
**Pull request by DanNixon**
Fixes #387

### CODE REVIEW

File: core/except.cpp [hunk 1]
```diff
@@ -305,5 +305,15 @@ void validExtent(const scipp::index size) {
     throw except::DimensionError("Dimension size cannot be negative.");
 }
 
+void binEdgeOverlaps(const VariableConstProxy &lhs,
+                     const VariableConstProxy &rhs, const Dim dim) {
+  const auto lhsLength = lhs.dims()[dim];
+
+  const auto lastBinOfLhs = lhs.slice({dim, lhsLength - 1, lhsLength});
+  const auto firstBinOfRhs = rhs.slice({dim, 0, 1});
```
Comment (SimonHeybrock) [inline core/except.cpp]: Is there a reason for slicing with a (size 1) range, instead of just using a single index?

File: core/except.cpp [hunk 2]
```diff
@@ -305,5 +305,15 @@ void validExtent(const scipp::index size) {
     throw except::DimensionError("Dimension size cannot be negative.");
 }
 
+void binEdgeOverlaps(const VariableConstProxy &lhs,
+                     const VariableConstProxy &rhs, const Dim dim) {
```
Comment (SimonHeybrock) [inline core/except.cpp]: I do not understand the function name. Isn't this checking that there is *no* overlap (but also no gap)?

File: core/dataset.cpp [hunk 3]
```diff
@@ -1184,4 +1191,75 @@ Dataset merge(const DatasetConstProxy &a, const DatasetConstProxy &b) {
...
+                        concatenate(a.coords()[dim],
+                                    b.coords()[dim].slice(
+                                        {dim, 1, b.coords()[dim].dims()[dim]}),
+                                    dim));
+        result.setData(std::string(name),
+                       concatenate(item.data(), b[name].data(), dim));
+      } else {
+        result.setCoord(dim,
+                        concatenate(a.coords()[dim], b.coords()[dim], dim));
+        result.setData(std::string(name),
```
Comment (SimonHeybrock) [inline core/dataset.cpp]: Move outside `if` to avoid duplication?

File: core/test/concatenate_test.cpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  Dataset b;
+  b.setCoord(Dim::X, makeVariable<int>({Dim::X, 3}, {4, 5, 6}));
+  b.setData("data_1", makeVariable<int>({Dim::X, 3}, {14, 15, 16}));
+
+  const auto d = concatenate(a, b, Dim::X);
+
+  EXPECT_EQ(makeVariable<int>({Dim::X, 6}, {1, 2, 3, 4, 5, 6}),
+            d.coords()[Dim::X]);
+  EXPECT_EQ(makeVariable<int>({Dim::X, 6}, {11, 12, 13, 14, 15, 16}),
+            d["data_1"].data());
```
Comment (SimonHeybrock) [inline core/test/concatenate_test.cpp]: Generally we are using `ASSERT(actual, expected)` throughout scipp (not the other way round), could you adopt that convention?

File: core/except.cpp (hunk 2 above)
Comment (DanNixon) [reply core/except.cpp]: Yes, maybe `binEdgeMatches` is a better name?

It's checking for an overlapping/matching bin edge, the fact it is edges not bins and edge is singular I assumed made it apparent is was only one.

Comment (SimonHeybrock) [reply core/except.cpp]: Still does not say what it does (could be comparing all bin edges?). Maybe `lastSliceEqualsFirstSlice`?

Comment (DanNixon) [reply core/except.cpp]: `lastSliceEqualsFirstSlice` doesn't indicate how big the slice is.

`binsAreConsectiveButNotOverlapping`?

Comment (SimonHeybrock) [reply core/except.cpp]: How about you just define this helper locally, then the name does not matter so much? Unless it is generic enough to be useful there is no need to put it inti `except.h`.

**Comment by SimonHeybrock** [comment]
Merging into a local branch to continue work.

**Pull request merged**